### PR TITLE
[camera] Fix for CameraAccessException that prevents image capture on certain devices running Android 7/8

### PR DESCRIPTION
**Pull request by carman247**
Fix for https://github.com/flutter/flutter/issues/88775 `CameraAccessException` that affects certain devices running Android 7/8 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by carman247** [comment]
Apologies if anything is incorrect/incomplete but this is my first PR.

It was suggested I create one for this bug. 

Any help would be greatly appreciated as although the changes I've made are minimal I've never done this before.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 1]
```diff
@@ -583,8 +583,12 @@ public void onCaptureCompleted(
         };
 
     try {
-      captureSession.stopRepeating();
-      captureSession.abortCaptures();
+      if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.O) {
+        captureSession.stopRepeating();
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:587]: Why is this line duplicated instead of outside of the if/else?

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 2]
```diff
@@ -583,8 +583,12 @@ public void onCaptureCompleted(
...
+      } else {
+        captureSession.stopRepeating();
+        captureSession.abortCaptures();
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:590]: There needs to be a comment explaining why this is conditional.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
> Apologies if anything is incorrect/incomplete but this is my first PR.

It mostly looks good, except for the item in the checklist that you missed; please see the linked documents about testing requirements.

**Comment by jgoyvaerts** [comment]
Actually, I'm pretty sure the captureSession.abortCaptures(); can be removed completely. I was about to make a PR for this myself and stumbled upon this issue.

According to [the documentation](https://developer.android.com/reference/android/hardware/camera2/CameraCaptureSession#abortCaptures())  it should only be called before a new session is created, however, calling .capture() does not create a new session. 

The .abortCaptures call also adds a significant delay between when the take picture button is pressed and when the shutter actually takes a picture (around 200ms on my Samsung S10), removing this line causes the delay of the shutter to dissappear.

I have not noticed any negative side effects from removing this line completely, git blame doesn't give any reason why it exists in the first place.

**Comment by carman247** [comment]
@jgoyvaerts Hi, so do you think the code can simply be `captureSession.stopRepeating();` and exclude the if statement altogether?

I'm also noticing some issues on specific devices (Alcatel 1C, and Pixel 6 Pro so far) that don't seem to be able to capture at all ...

I'm wondering if it's something related to the `.abortCaptures();` ?

**Comment by jgoyvaerts** [comment]
I'm just saying based on my tests, the .abortCaptures call 
- Has no effect on being able to take a picture on my device, I can take a picture just fine when the call is removed
- Introduces an arbitrary delay of 150-200ms when taking a picture
- The documentation doesn't say why this should be called before taking a picture, only when you want to abort a picture capture process in progress (like a long shutter exposure or something).

Based on these reasons, I'm inclined to believe the line can be removed, especially if you have noticed it causes issues on certain Android versions.

Maybe there's an edge case we're missing why it needs to be there, but then that edge case should probably be fixed in a different way, since the happy flow of taking a picture should not have an unneccessary 150-200ms delay.

**Comment by carman247** [comment]
@jgoyvaerts Ok. I will test on the three different devices I have, and potentially test on some cloud devices as well.

However, just to confirm. I'm not sure at all if this `abortCaptures` is the cause of the problems on the devices I mentioned.

If I find no problems removing it then I'm tempted to test it out in the wild although certainly not the best way to test whether it's worked or not.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java (hunk 2 above)
Comment (subzero911) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:589]: I can confirm that this line causes troubles on some old Huawei/Xiaomi devices.

**Comment by jgoyvaerts** [comment]
@carman247 Can you just adjust the PR so it removes the line captureSession.abortCaptures(); completely?

**Comment by carman247** [comment]
@jgoyvaerts no problem, will get that updated in a moment.

**Comment by carman247** [comment]
Can anyone explain why the checks are failing? Not sure what I'm doing wrong there.

Also, I haven't added any new "tests" for this PR .. I'm assuming the current ones will still be fine.

**Comment by stuartmorgan** [comment]
> Can anyone explain why the checks are failing? Not sure what I'm doing wrong there.

The repository keys are rotated periodically; you likely just need to merge in the latest changes from `main`.

> Also, I haven't added any new "tests" for this PR .. I'm assuming the current ones will still be fine.

This PR doesn't meet any of the test exemption criteria (see the link in that checkbox item), so this needs tests. See also the comment from the bot above. If having this call breaks image capture on certain devices, then there need to be a test that the call isn't being made in this flow.

**Comment by jgoyvaerts** [comment]
> This PR doesn't meet any of the test exemption criteria (see the link in that checkbox item), so this needs tests. See also the comment from the bot above. If having this call breaks image capture on certain devices, then there need to be a test that the call isn't being made in this flow.
The PR has been changed to completely delete the line, since it serves no purpose AND causes issues on certain devices. 

I thought delete only PRs did not require tests?

**Comment by stuartmorgan** [comment]
> I thought delete only PRs did not require tests?

I missed that it was delete-only now; my comment was based on the previous state.

That said, this should still have a test, for the reason described in the bot comment above: someone added this once, thinking it was a good idea. If they do so again, do you want all of the currently affected devices to silently regress again?

**Comment by stuartmorgan** [comment]
@mvanbeusekom You added this line in https://github.com/flutter/plugins/pull/3383; do you remember why by any chance?

@carman247 Have you verified that this doesn't regress the fix in that PR?

**Comment by carman247** [comment]
> @carman247 Have you verified that this doesn't regress the fix in that PR?

@stuartmorgan I'll check and confirm that asap

**Comment by stuartmorgan** [comment]
> > You added this line in #3383; do you remember why by any chance?
> 
> @stuartmorgan It looks like `abortCaptures` was added to `startVideoRecording` in that PR, and not `takePictureAfterPrecapture`

Whoops, you're right; looks like my git query was wrong. It's actually from one of the shards of https://github.com/flutter/plugins/pull/3651

@acoutts Do you know why that line is there?

**Comment by stuartmorgan** [comment]
On the testing side: I would expect it would be straightforward to add a test to https://github.com/flutter/plugins/blob/main/packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/CameraTest.java that uses the mock capture session to ensure that this isn't being called.

**Comment by stuartmorgan** [comment]
@carman247 Are you planning on adding tests to this so it can move forward?

**Comment by carman247** [comment]
> @carman247 Are you planning on adding tests to this so it can move forward?

Hi, yes and sorry. I'm very busy recently and don't have a lot of free time that I can spend on much else. I'm also not very well versed in Java and have never written/used tests in java. If anyone else is able to help or write the test that would help me a lot!

**Comment by stuartmorgan** [comment]
> On the testing side: I would expect it would be straightforward to add a test to https://github.com/flutter/plugins/blob/main/packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/CameraTest.java that uses the mock capture session to ensure that this isn't being called.

I took a look at this in more detail, and was not as straightforward as I thought since `CameraDevice` is not easily mockable since it's partially abstract and has a private constructor. I inserted a wrapper that allows mocking it, and tested using that.

@blasten Please review closely, since I'm very out of practice with Java.

**Comment by stuartmorgan** [comment]
@BeMacized Could you also take a look since you were in this code quite a bit?

**Comment by carman247** [comment]
@stuartmorgan thanks for sorting the test! Glad you tackled it as would've taken me ages to get in done and would probably be wrong anyway 😅

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 3]
```diff
@@ -79,6 +79,24 @@
   void onError(String errorCode, String errorMessage);
 }
 
+/** A mockable wrapper for CameraDevice calls. */
+interface CameraDeviceWrapper {
+  @NonNull
+  CaptureRequest.Builder createCaptureRequest(int var1) throws CameraAccessException;
```
Comment (blasten) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:85]: var1 doesn't seem to provide meaning. Did you mean templateType?

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 4]
```diff
@@ -79,6 +79,24 @@
   void onError(String errorCode, String errorMessage);
 }
 
+/** A mockable wrapper for CameraDevice calls. */
+interface CameraDeviceWrapper {
```
Comment (blasten) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:83]: I think this is a Delegate. `CameraDeviceDelegate` works.

https://en.wikipedia.org/wiki/Delegation_pattern

Comment (stuartmorgan) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:83]: I don't see how this is a delegate; nothing is being delegated to it. It's just an interface to allow mocking a `CameraDevice`

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java (hunk 3 above)
Comment (stuartmorgan) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java]: Oops! The declarations of these methods didn't have useful variable names for some reason. I thought I fixed them all when I copied them.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java (hunk 4 above)
Comment (blasten) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:83]: oh I see, DefaultCameraDeviceWrapper delegates to CameraDevice

**Review by blasten — APPROVED** [review]
LGTM

**Review by stuartmorgan — APPROVED** [review]
Explicitly approving the non-test parts (I.e., the parts I didn't write) for the bot.
